x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java

```diff
@@ -259,38 +262,68 @@ private void ensureOpen() {
         }
     }
 
-    CompletableFuture<Integer> fetchRange(
-        long start,
-        long end,
-        CheckedBiFunction<Long, Long, Integer, IOException> onRangeAvailable,
-        CheckedBiConsumer<Long, Long, IOException> onRangeMissing
+    @FunctionalInterface
+    interface CacheReader {
+        int read(FileChannel channel) throws IOException;
+    }
+
+    @FunctionalInterface
+    interface CacheWriter {
+        void write(FileChannel channel, long from, long to, Consumer<Long> progressUpdater) throws IOException;
```

Comment:
What, no `CheckedQuadConsumer<...>`? 😁 

I'm torn about having these be separate interfaces vs combining them together perhaps along with some other arguments to `fetchAsync` too. I'm finding it especially strange that `CacheReader#read` only takes a `FileChannel`; this indicates that `rangeToRead` is now available, but that's very implicit now.

Maybe different names would help. How about something like `RangeMissingHandler#fillCacheRange` and `RangeAvailableHandler#onRangeAvailable`?

Reply:
> What, no CheckedQuadConsumer<...>? grin

Don't tempt me :)



> Maybe different names would help. How about something like `RangeMissingHandler#fillCacheRange` and `RangeAvailableHandler#onRangeAvailable`?

I went this route and renamed to RangeMissingHandler/RangeAvailableHandler